ledger/ledger-api-client/src/it/scala/com/digitalasset/ledger/client/UpgradesIT.scala

```diff
@@ -0,0 +1,156 @@
+        case Some(failureMessage) => {
+          cantonLog should include regex s"Typechecking upgrades for $testPackageV2Id failed with following message: $failureMessage"
+          uploadV2Result match {
+            case None => fail(s"Uploading second package $testPackageV2Id should fail but didn't.");
+            case Some(err) => {
+              val msg = err.toString
+              msg should include regex "INVALID_ARGUMENT: DAR_NOT_VALID_UPGRADE"
+              msg should include regex "The DAR contains a package which claims to upgrade another package, but basic checks indicate the package is not a valid upgrade"
+            }
+          }
+        }
+
+        // If a failure is not expected, look for a success message
+        case None => {
+          cantonLog should include regex s"Typechecking upgrades for $testPackageV2Id succeeded."
+          uploadV2Result match {
+            case None => succeed;
+            case Some(err) => {
+              fail(
+                s"Uploading second package $testPackageV2Id shouldn't fail but did, with message: $err"
+              );
+            }
+          }
+        }
+      }
+    }
+  }
+
+  "Upload-time Upgradeability Checks" should {
+    "report no upgrade errors for valid upgrade" in {
+      testPackagePair(
+        "test-common/upgrades-ValidUpgrade-v1.dar",
+        "test-common/upgrades-ValidUpgrade-v2.dar",
+        None,
+      )
+    }
+    "report error when module is missing in upgrading package" in {
+      testPackagePair(
+        "test-common/upgrades-MissingModule-v1.dar",
+        "test-common/upgrades-MissingModule-v2.dar",
+        Some("MissingModule"),
+      )
+    }
+    "report error when template is missing in upgrading package" in {
+      testPackagePair(
+        "test-common/upgrades-MissingTemplate-v1.dar",
+        "test-common/upgrades-MissingTemplate-v2.dar",
+        Some("MissingTemplate"),
+      )
+    }
+    "report error when datatype is missing in upgrading package" in {
+      testPackagePair(
+        "test-common/upgrades-MissingDataCon-v1.dar",
+        "test-common/upgrades-MissingDataCon-v2.dar",
+        Some("MissingDataCon"),
+      )
+    }
+    "report error when choice is missing in upgrading package" in {
+      testPackagePair(
+        "test-common/upgrades-MissingChoice-v1.dar",
+        "test-common/upgrades-MissingChoice-v2.dar",
+        Some("MissingChoice"),
+      )
+    }
+    "report error when key type changes" in {
+      testPackagePair(
+        "test-common/upgrades-TemplateChangedKeyType-v1.dar",
+        "test-common/upgrades-TemplateChangedKeyType-v2.dar",
+        Some("TemplateChangedKeyType"),
+      )
+    }
+    "report error when record fields change" in {
+      testPackagePair(
+        "test-common/upgrades-RecordFieldsNewNonOptional-v1.dar",
+        "test-common/upgrades-RecordFieldsNewNonOptional-v2.dar",
+        Some("RecordFieldsNewNonOptional"),
+      )
+    }
+  }
```

Comment:
Is there anything more to test? feels like there should be more cases given the logic behind them, but I'm not sure if thats the case

Reply:
DamlcUpgrades.hs is a good reference, will add more tests once the variants checking is implemented

Reply:
I've copied over all the tests from DamlcUpgrades, fixed